### ISSUE TITLE
FileNotFoundException with gRRC_for_.NET on Unity when using NuGetForUnity

### ISSUE THREAD

**Issue by m-yukio** [protocolbuffers/protobuf#9618]
FileNotFoundException with gRRC_for_.NET on Unity when using NuGetForUnity

**What version of protobuf and what language are you using?**
Version: v3.19.4
Language: C#
Package: NuGet for Unity

**What operating system (Linux, Windows, ...) and version?**
macOS Monterey 12.2.1

**What runtime / compiler are you using (e.g., python version or gcc version)**
Unity 2021.2.11f1

**What did you do?**
1. Install the package with NuGet for .NET.
  - Google.Protobuf
  - Grpc.Net.Client
```
.-- Assets
    |-- NuGet
    `-- Packages
        |-- Google.Protobuf.3.19.4/lib/netstandard2.0/Google.Protobuf.dll
        |-- Grpc.Core.Api.2.43.0/lib/netstandard2.0/Grpc.Core.Api.dll
        |-- Grpc.Net.Client.2.43.0/lib/netstandard2.0/Grpc.Net.Client.dll
        |-- Grpc.Net.Common.2.43.0/lib/netstandard2.0/Grpc.Net.Common.dll
        |-- Microsoft.Extensions.Logging.Abstractions.3.0.3/lib/netstandard2.0/Microsoft.Extensions.Logging.Abstractions.dll
        `-- System.Diagnostics.DiagnosticSource.4.5.1/lib/netstandard1.3/System.Diagnostics.DiagnosticSource.dll
```
2. Coding the gRPC client with .NET Standard 2.0
3. Launch a gRPC server
4. Communicate with the server
5. See error

```
RpcException: Status(StatusCode="Unavailable", Detail="Error starting gRPC call. HttpRequestException: Error while copying content to a stream. FileNotFoundException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.", DebugException="System.Net.Http.HttpRequestException: Error while copying content to a stream. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
  at Google.Protobuf.WritingPrimitives.WriteString (System.Span`1[System.Byte]& buffer, Google.Protobuf.WriterInternalState& state, System.String value) [0x00076] in <2ccdeadc259d431f8c7eae367a5d08d1>:0 
  at Google.Protobuf.WriteContext.WriteString (System.String value) [0x00000] in <2ccdeadc259d431f8c7eae367a5d08d1>:0 
  at Helloworld.HelloRequest.pb::Google.Protobuf.IBufferMessage.InternalWriteTo (Google.Protobuf.WriteContext& output) [0x0001d] in /Users/murakami_yukio/Projects/murakami_yukio/Stubby/Stubby/Assets/Scripts/Proto/Hello.cs:184 
  at Google.Protobuf.WritingPrimitivesMessages.WriteRawMessage (Google.Protobuf.WriteContext& ctx, Google.Protobuf.IMessage message) [0x0000a] in <2ccdeadc259d431f8c7eae367a5d08d1>:0 
  at Google.Protobuf.MessageExtensions.WriteTo (Google.Protobuf.IMessage message, System.Buffers.IBufferWriter`1[T] output) [0x00020] in <2ccdeadc259d431f8c7eae367a5d08d1>:0 
  at Helloworld.Greeter.__Helper_SerializeMessage (Google.Protobuf.IMessage message, Grpc.Core.SerializationContext context) [0x0001c] in /Users/murakami_yukio/Projects/murakami_yukio/Stubby/Stubby/Assets/Scripts/Greeter.cs:26 
  at Grpc.Net.Client.StreamExtensions.WriteMessageAsync[TMessage] (System.IO.Stream stream, Grpc.Net.Client.Internal.GrpcCall call, TMessage message, System.Action`2[T1,T2] serializer, Grpc.Core.CallOptions callOptions) [0x00108] in <93d755b016cc4513ac20ab2598098c88>:0 
  at Grpc.Net.Client.Internal.PushUnaryContent`2[TRequest,TResponse].WriteMessageCore (System.Threading.Tasks.Task writeMessageTask) [0x00064] in <93d755b016cc4513ac20ab2598098c88>:0 
  at System.Net.Http.HttpContent.CopyToAsyncCore (System.Threading.Tasks.ValueTask copyTask) [0x00067] in <156582be1ebd4171b9d70018d8f80e4f>:0 
   --- End of inner exception stack trace ---
  at System.Net.Http.HttpContent.CopyToAsyncCore (System.Threading.Tasks.ValueTask copyTask) [0x0008f] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at System.Net.Http.HttpConnection.SendRequestContentAsync (System.Net.Http.HttpRequestMessage request, System.Net.Http.HttpConnection+HttpContentWriteStream stream, System.Threading.CancellationToken cancellationToken) [0x000a0] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at System.Net.Http.HttpConnection.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x012e6] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync (System.Net.Http.HttpConnection connection, System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0012b] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at System.Net.Http.HttpConnectionPool.SendWithRetryAsync (System.Net.Http.HttpRequestMessage request, System.Boolean doRequestAuth, System.Threading.CancellationToken cancellationToken) [0x0014b] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at System.Net.Http.RedirectHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x000ba] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered (System.Threading.Tasks.Task`1[TResult] sendTask, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationTokenSource cts, System.Boolean disposeCts) [0x000b3] in <156582be1ebd4171b9d70018d8f80e4f>:0 
  at Grpc.Net.Client.Internal.GrpcCall`2[TRequest,TResponse].RunCall (System.Net.Http.HttpRequestMessage request, System.Nullable`1[T] timeout) [0x00215] in <93d755b016cc4513ac20ab2598098c88>:0 ")
Grpc.Net.Client.Internal.HttpClientCallInvoker.BlockingUnaryCall[TRequest,TResponse] (Grpc.Core.Method`2[TRequest,TResponse] method, System.String host, Grpc.Core.CallOptions options, TRequest request) (at <93d755b016cc4513ac20ab2598098c88>:0)
Grpc.Core.Interceptors.InterceptingCallInvoker.<BlockingUnaryCall>b__3_0[TRequest,TResponse] (TRequest req, Grpc.Core.Interceptors.ClientInterceptorContext`2[TRequest,TResponse] ctx) (at <4f08a803638948218f0b2255b55def9b>:0)
Grpc.Core.ClientBase+ClientBaseConfiguration+ClientBaseConfigurationInterceptor.BlockingUnaryCall[TRequest,TResponse] (TRequest request, Grpc.Core.Interceptors.ClientInterceptorContext`2[TRequest,TResponse] context, Grpc.Core.Interceptors.Interceptor+BlockingUnaryCallContinuation`2[TRequest,TResponse] continuation) (at <4f08a803638948218f0b2255b55def9b>:0)
Grpc.Core.Interceptors.InterceptingCallInvoker.BlockingUnaryCall[TRequest,TResponse] (Grpc.Core.Method`2[TRequest,TResponse] method, System.String host, Grpc.Core.CallOptions options, TRequest request) (at <4f08a803638948218f0b2255b55def9b>:0)
Helloworld.Greeter+GreeterClient.SayHello (Helloworld.HelloRequest request, Grpc.Core.CallOptions options) (at Assets/Scripts/Greeter.cs:144)
Helloworld.Greeter+GreeterClient.SayHello (Helloworld.HelloRequest request, Grpc.Core.Metadata headers, System.Nullable`1[T] deadline, System.Threading.CancellationToken cancellationToken) (at Assets/Scripts/Greeter.cs:133)
GrpcTest.RunHelloWorld (System.Boolean useNet) (at Assets/Scripts/GrpcTest.cs:82)
GrpcTest.RunNet () (at Assets/Scripts/GrpcTest.cs:34)
UnityEngine.Events.InvokableCall.Invoke () (at /Users/bokken/buildslave/unity/build/Runtime/Export/UnityEvent/UnityEvent.cs:178)
UnityEngine.Events.UnityEvent.Invoke () (at /Users/bokken/buildslave/unity/build/Runtime/Export/UnityEvent/UnityEvent/UnityEvent_0.cs:58)
UnityEngine.UI.Button.Press () (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/UI/Core/Button.cs:70)
UnityEngine.UI.Button.OnPointerClick (UnityEngine.EventSystems.PointerEventData eventData) (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/UI/Core/Button.cs:114)
UnityEngine.EventSystems.ExecuteEvents.Execute (UnityEngine.EventSystems.IPointerClickHandler handler, UnityEngine.EventSystems.BaseEventData eventData) (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/ExecuteEvents.cs:57)
UnityEngine.EventSystems.ExecuteEvents.Execute[T] (UnityEngine.GameObject target, UnityEngine.EventSystems.BaseEventData eventData, UnityEngine.EventSystems.ExecuteEvents+EventFunction`1[T1] functor) (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/ExecuteEvents.cs:272)
UnityEngine.EventSystems.EventSystem:Update() (at Library/PackageCache/com.unity.ugui@1.0.0/Runtime/EventSystem/EventSystem.cs:501)
```

**What did you expect to see**
**What did you see instead?**

When I tried to output the log on the gRPC server, it seems that the connection is established.

```
2022/03/11 11:37:55 server listening at 192.168.200.3:50051
2022/03/11 11:38:08 http2: Framer 0xc0001f2000: wrote SETTINGS len=6, settings: MAX_FRAME_SIZE=16384
```

**Anything else we should know about your project / environment**

Using gRPC C# core library instead of gRPC for .NET will succeed.
Even in gRPC C# core library, protobuf is a C # DLL.

```
.-- Assets
    `-- Plubins
        |-- Google.Protobuf/lib/net45/Google.Protobuf.dll
        |-- Grpc.Core
        |   |-- lib/net45/Grpc.Core.dll
        |   `-- runtimes/<platform>/<arch>/libgrpc_csharp_ext.so etc
        |-- Grpc.Core.Api/lib/net45/Grpc.Core.Api.dll
        |-- System.Buffers/lib/net45/System.Buffers.dll
        |-- System.Memory/lib/net45/System.Memory.dll
        `-- System.Runtime.CompilerServices.Unsafe/lib/net45/System.Runtime.CompilerServices.Unsafe.dll
```

**Comment by amanda-tarafa** [comment]
[Starting on Unity 2020.1, the System.Runtime.CompilerServices.Unsafe.dll was removed from Unity.Collections package](https://forum.unity.com/threads/please-update-com-unity-collections-reference-of-system-runtime-compilerservices-unsafe.732056/#post-5972735). Basically the package is not present on Unity's .NET implementation (by default).

You might be able to make it work by providing the .dll yourself, note how that's precisely what's happening when you try your code with Grpc.Core.

There's not much else we can do on this repo, as this is just Unity moving away from .NET standards, which will continue to follow.

**Comment by amanda-tarafa** [comment]
As per my last comment, I'll close this issue now, but do feel free to leave a comment if you belive we could help further.

**Comment by tayler6000** [comment]
Forgive me for commenting on an old thread Maintainers, but I wanted to leave an extra breadcrumb for the next person.  The recommendation of bringing over the missing DLL worked for me.  My Unity is currently using .NET 2.1, so I downloaded the binaries from Microsoft.  Unfortunately, the .Net 2.1 binaries download did not have the necessary file.  I tried .Net 8 (specifically .8 SDK 8.0.300-win-x86) hoping that this file hasn't changed much and everything successfully compiled.  Haven't tried using the protobufs yet, but compiling is a good sign and I'll edit this post if it turns out it didn't work. When you download the zip the file will be in `shared/Microsoft.NETCore.App/8.0.5/System.Runtime.CompilerServices.Unsafe.dll`. Good luck!
